ros2bag/ros2bag/reindexer/bag_metadata.py

```diff
@@ -0,0 +1,304 @@
+# Copyright 2020 DCS Corporation, All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# DISTRIBUTION A. Approved for public release; distribution unlimited.
+# OPSEC #4584.
+#
+# Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
+# Part 252.227-7013 or 7014 (Feb 2014).
+#
+# This notice must appear in all copies of this file and its derivatives.
+
+import pathlib
+from typing import Dict, List, Literal, TypedDict, Union
+
+from ros2bag.api import print_error
+import yaml
+
+
+class TopicBlock(TypedDict):
+    block_name: str
+    block_type: str
+    block_ser: str
+    block_qos: str
+
+
+def block_as_dict(topic_block: TopicBlock) -> Dict[str, str]:
+    return {
+        'name': topic_block['block_name'],
+        'type': topic_block['block_type'],
+        'serialization_format': topic_block['block_ser'],
+        'offered_qos_profiles': topic_block['block_qos']
+    }
+
+
+class TopicMetadata(TypedDict):
+    """Contains the metadata for a ROS 2 topic."""
+
+    tm_block: TopicBlock
+    tm_message_count: int
+
+
+def topic_as_dict(topic: TopicMetadata) -> Dict[str, Union[Dict[str, str], str, int]]:
+    metadata_topics = block_as_dict(topic['tm_block'])
+    return {'topic_metadata': metadata_topics, 'message_count': topic['tm_message_count']}
+
+
+class RosbagYamlDumper(yaml.Dumper):
+
+    def increase_indent(self, flow=False, indentless=False):
+        return super(RosbagYamlDumper, self).increase_indent(flow, False)
+
+
+class MetadataWriter:
```

Comment:
Considering we already have a `MetadataWriter` in C++, this duplicated code will be prone to rot if the metadata format changes at all - this should be a call into rosbag2_storage `MetadataIO` for the logic in this file

Reply:
I don't think I've seen ROS code in the past calling C++ libraries from within Python code. Is there precedent for this, or is this something that will change going forward?